keras_cv/losses/dice.py

```diff
@@ -0,0 +1,215 @@
+
+      Standalone usage:
+
+      >>> y_true = tf.random.uniform([5, 10, 10], 0, maxval=4, dtype=tf.int32)
+      >>> y_true = tf.one_hot(y_true, depth=4)
+      >>> y_pred = tf.random.uniform([5, 10, 10, 4], 0, maxval=4)
+      >>> dice = Dice()
+      >>> dice(y_true, y_pred).numpy()
+      0.5549314
+
+      >>> # Calling with 'sample_weight'.
+      >>> dice(y_true, y_pred, sample_weight=tf.constant([[0.5, 0.5]])).numpy()
+      0.24619368
+
+      Usage with the `compile()` API:
+      ```python
+      model.compile(optimizer='adam', loss=keras_cv.losses.Dice())
+    ```
+    """
+
+    def __init__(
+        self,
+        beta=1,
+        from_logits=False,
+        class_ids=None,
+        axis=[1, 2],
+        loss_type=None,
+        label_smoothing=0.0,
+        per_sample=False,
+        epsilon=1e-07,
+        name="dice",
+        **kwargs,
+    ):
+        """Initializes `Dice` instance.
+
+        Args:
+            beta: A float or integer coefficient for balancing the precision
+                and recall. It determines the weight of recall and precision
+                in the combined score. If `beta < 1`, precisoin will doninate;
+                if `beta > 1`, recall will dominate. Default to `1`.
+            from_logits: Whether `y_pred` is expected to be a logits tensor. By
+                default, we assume that `y_pred` encodes a probability distribution.
+                Default to `False`.
+            class_ids: An interger or a lost of intergers within `range(num_classes)`
+                to evaluate the loss. If it's `None`, all classes will beu used to
+                calculate the loss. Default to `None`.
+            axis: An optional sequence of `int` specifying the axis to perform reduce
+                ops for raw dice score. For 2D model, it should be [1,2] or [2,3] for
+                the `channels_last` or `channels_first` format respectively. And for
+                3D mdoel, it should be [1,2,3] or [2,3,4] for the `channels_last` or
+                `channel_first` format respectively.
+            loss_type: An optional `str` specifying the type of the dice score to
+                compute. Compute generalized or adaptive dice score if loss type is
+                `generalized` or `adaptive`; otherwise compute original dice score.
+                Default to `None`.
+            label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+                meaning the confidence on label values are relaxed. For example, if
+                `0.1`, use `0.1 / num_classes` for non-target labels and
+                `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+            per_sample: If `True`, the loss will be calculated for each sample in
+                batch and then averaged. Otherwise the loss will be calculated for
+                the whole batch. Default to `False`.
+            epsilon: Small float added to dice score to avoid dividing by zero.
```

Comment:
I think this one is probably not needed. We can just use keras.backend.epsilon() directly.

Reply:
This one is the smoothing parameter and sometimes users may want to set their own. WDYT? But it can be removed of course. As a default value, I use `backend.epsilon` though.